### PR TITLE
fix(ai): 修复 BYOP tool result 晚到乱序

### PR DESCRIPTION
**Pull request by hemo94931**
按 params.input 原顺序持久化 UserQuery/ActionResult，避免用户继续输入时形成 Assistant(tool_call) -> UserQuery -> ToolCallResult 的历史。

sanitize_tool_call_pairs 现在按 call_id 重建 ToolResponse，真实结果会覆盖 placeholder；同时保留 streamed tool call 顺序并补充 byop-diag 日志和回归测试。

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved handling and ordering consistency of concurrent tool calls
  * Enhanced stability when tool responses arrive out-of-order
  * Strengthened diagnostic logging for request troubleshooting

* **Tests**
  * Extended coverage for late tool responses and edge case scenarios

[Review Change Stack](https://app.coderabbit.ai/change-stack/zerx-lab/warp/pull/96?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: app/src/ai/agent_providers/chat_stream.rs [hunk 1]
```diff
@@ -1140,62 +1441,85 @@ fn apply_caching_anthropic(messages: &mut Vec<ChatMessage>) {
...
-        }
+        rewritten.push(ChatMessage::from(bundled));
+    }
 
-        // 用合并后的单条 Tool message 替换 [i+1..j) 这段原始连续 Tool messages。
-        messages.splice(i + 1..j, std::iter::once(ChatMessage::from(bundled)));
-        i += 2; // 跳过 Assistant + 合并后的 Tool message
+    if !responses_by_call_id.is_empty() {
+        orphan_call_ids.extend(responses_by_call_id.into_keys());
     }
```
Comment (zerx-lab) [inline app/src/ai/agent_providers/chat_stream.rs:1521]: **【缓存命中 — 高优先级风险】sanitize 跨位置重排会破坏 Anthropic prompt cache prefix 稳定性**

新实现把所有 Tool message 提取后**统一从历史中删除**(`for msg in messages.drain(..) { if msg.role == Tool { continue; } }`),再把真实 result 重新塞回对应 Assistant 之后。

这在功能上修复了 bug,但**改变了 messages 序列的物理位置**:
- 旧历史:`[..., Assistant(tool_call=a), User("打断"), Tool(real_a)]`
- 新历史:`[..., Assistant(tool_call=a), Tool(real_a), User("打断")]`

下一轮请求,`apply_caching_anthropic`(L1442+)给"last 2 non-system message"打 5m breakpoint。如果这种重排发生在历史尾部附近,则:
1. 同一个对话第 N+1 轮重建 messages 时,该段的字节序与第 N 轮**完全不同**;
2. 上游 Anthropic 按 prefix 哈希查 cache → **必然 miss**,触发一次 5m / 1h cache write(`2x base`)。
3. 后续每轮只要这段历史还出现在末尾两条,都会因相对位置漂移而反复 cache write。

更糟的是,**当前轮**的 sanitize 也会在多 tool_call 场景下触发重排,Anthropic 的 `tool_result` block 排序要求虽满足了,但 cache breakpoint 位置在每轮可能跳。

**建议**:
1. 在函数末尾加日志(类似 placeholders/orphans),统计每次调用是否实际发生了"跨位置回搬"事件,便于在 `[byop-cache]` log 里关联 miss ratio 突增。
2. 评估是否在"竞态历史"被检测到后,顺手用 `record_event` 之类标识本轮 cache 必然 miss,避免误判 P0-4/P0-5 失效。
3. 文档里(本函数 doc comment 第二段)显式声明"跨位置回搬会破坏 prefix 一致性",未来 cache 命中率分析脚本(`script/analyze-prompt-cache.ps1`)需把此函数 warn 行作为 miss 归因之一。

**注**:这只是 *可观测性* 层面的建议,功能本身正确;但与原 doc comment "不做跨位置重排 Tool messages" 的承诺**明显矛盾**,doc 必须同步更新。

File: app/src/ai/agent_providers/chat_stream.rs [hunk 2]
```diff
@@ -1140,62 +1441,85 @@ fn apply_caching_anthropic(messages: &mut Vec<ChatMessage>) {
...
-        while j < messages.len() && messages[j].role == genai::chat::ChatRole::Tool {
-            for resp in messages[j].content.tool_responses() {
-                existing.insert(resp.call_id.clone(), (*resp).clone());
-            }
-            j += 1;
+        let expected_call_ids: Vec<String> = if msg.role == genai::chat::ChatRole::Assistant {
+            msg.content
+                .tool_calls()
+                .iter()
+                .map(|tc| tc.call_id.clone())
```
Comment (zerx-lab) [inline app/src/ai/agent_providers/chat_stream.rs:1493]: **【正确性 — 中优先级】placeholder 覆盖语义需要更严谨**

```rust
match responses_by_call_id.get(&resp.call_id) {
    Some(existing)
        if !is_placeholder_response(resp) || is_placeholder_response(existing) =>
    {
        responses_by_call_id.insert(resp.call_id.clone(), (*resp).clone());
    }
    None => { ... insert ... }
    Some(_) => {}
}
```

这段逻辑想表达:
- 同 call_id 现有 placeholder + 新真实 result → 覆盖 ✅
- 同 call_id 现有真实 result + 新 placeholder → 不覆盖 ✅
- 同 call_id 现有真实 result + 新真实 result → 后者覆盖(`!is_placeholder(resp)` 成立)✅
- 同 call_id 现有 placeholder + 新 placeholder → 不覆盖(都是占位,后者也无意义)❌ —— 实际上当前会**走 None=>insert 分支吗?不,走 `Some(existing) if !is_placeholder(resp) || is_placeholder(existing)`,因为 existing 是 placeholder,条件 `is_placeholder(existing) = true` 触发 OR 短路,会覆盖。

也就是说目前 placeholder→placeholder 也会覆盖一次,**功能等价**(都是同一占位文本),但增加无谓 clone。

更重要的是:**真实 result A → 真实 result B**(都不是 placeholder)情况下,后者会覆盖前者 —— 这是合理的"后到为准"语义,但**没有 unit test 覆盖**(测试只有"placeholder ← real"一种方向)。

**建议**:
1. 增加测试 `real_tool_response_is_replaced_by_later_real_tool_response`,验证多次真实 result 时取最新值(对齐旧实现的 `insert` 后到覆盖语义)。
2. 简化条件为更易读的形式:
   ```rust
   let should_replace = match responses_by_call_id.get(&resp.call_id) {
       None => true,
       Some(existing) => is_placeholder_response(existing) || !is_placeholder_response(resp),
   };
   if should_replace { responses_by_call_id.insert(...); }
   ```
   语义等价但意图更直白。

File: app/src/ai/agent_providers/chat_stream.rs [hunk 3]
```diff
@@ -2637,8 +2972,20 @@ pub async fn generate_byop_output(
...
                         for call in content.tool_calls() {
-                            tool_bufs.entry(call.call_id.clone()).or_insert_with(|| call.clone());
+                            if !captured_order.contains(&call.call_id) {
+                                captured_order.push(call.call_id.clone());
+                            }
+                            tool_bufs.insert(call.call_id.clone(), call.clone());
+                        }
+                        if !captured_order.is_empty() {
+                            for call_id in &tool_order {
+                                if !captured_order.contains(call_id) {
```
Comment (zerx-lab) [inline app/src/ai/agent_providers/chat_stream.rs:2984]: **【正确性 — 中优先级】End 处理对 `tool_order` 的"captured 优先"合并有潜在乱序**

新逻辑:
```rust
if let Some(content) = end.captured_content.as_ref() {
    let mut captured_order: Vec<String> = Vec::new();
    for call in content.tool_calls() {
        if !captured_order.contains(&call.call_id) {
            captured_order.push(call.call_id.clone());
        }
        tool_bufs.insert(call.call_id.clone(), call.clone());  // ← 注意:无脑覆盖
    }
    if !captured_order.is_empty() {
        for call_id in &tool_order {
            if !captured_order.contains(call_id) {
                captured_order.push(call_id.clone());
            }
        }
        tool_order = captured_order;
    }
}
```

两个问题:

**1. `tool_bufs.insert(... call.clone())` 改变了旧行为**

旧实现是 `tool_bufs.entry(...).or_insert_with(|| call.clone())`,即"streaming 已有 → 保留 streaming 版本"。新实现改成 `insert`(无条件覆盖) —— 注释还写着"优先用 captured_content 里的 tool_calls(更完整)",但旧代码的 `or_insert` 实际上是相反语义。

这是有意修改吗?如果是,需要在 PR 描述里说明;如果不是,会导致**前面 ToolCallChunk 累积的 args 增量被 captured 的最终版本覆盖** —— 在多数 provider 上 captured 才是完整的,**功能上更对**,但已经 emit 的 placeholder UI 卡片可能与最终 args 不同(因为 placeholder 用了 streaming 中间态的 parse 结果)。不算严重。

**2. `captured_order` 用 `Vec::contains` 做去重,O(N²)**

`captured_order.contains(&call.call_id)` 与最后的 `for call_id in &tool_order { if !captured_order.contains(call_id) }` 都是线性扫描。多 tool 并发场景 N 通常 ≤ 20,问题不大;但配上前面 ChunkArgs 阶段也有 `if !tool_bufs.contains_key(&call.call_id) { tool_order.push(...) }`,流末两层 O(N²) 累计。

**建议**:用 `HashSet<String>` 做存在性判断,或在 `tool_order` 本身用 `IndexMap` 替代 HashMap+Vec 的双轨结构。当前实现可接受,但若以后单轮 tool 数上千会成瓶颈。

File: app/src/ai/agent_providers/chat_stream.rs [hunk 4]
```diff
@@ -2728,7 +3075,16 @@ pub async fn generate_byop_output(
...
+        let mut ordered_tool_calls: Vec<ToolCall> = Vec::with_capacity(tool_bufs.len());
+        for call_id in tool_order {
+            if let Some(call) = tool_bufs.remove(&call_id) {
+                ordered_tool_calls.push(call);
+            }
+        }
+        let mut unordered_tool_calls: Vec<ToolCall> = tool_bufs.into_values().collect();
+        unordered_tool_calls.sort_by(|a, b| a.call_id.cmp(&b.call_id));
+        ordered_tool_calls.extend(unordered_tool_calls);
+        for call in ordered_tool_calls {
```
Comment (zerx-lab) [inline app/src/ai/agent_providers/chat_stream.rs:3087]: **【边缘情况】流末 tool_calls 排序的"剩余 unordered 按 call_id 字典序"是个安全网,但要保证不被踩到**

```rust
let mut unordered_tool_calls: Vec<ToolCall> = tool_bufs.into_values().collect();
unordered_tool_calls.sort_by(|a, b| a.call_id.cmp(&b.call_id));
ordered_tool_calls.extend(unordered_tool_calls);
```

理论上 `tool_bufs` 在 ChunkArgs / End 两条路径上都同步维护了 `tool_order`,流末走到这里时 `tool_bufs` **应该是空的**。这段 fallback 只在以下异常情况触发:
1. captured_content 里有 tool_call,但 ChunkArgs 阶段没见过(provider 不发增量 chunk,只在 End emit 完整工具调用)。但 End 分支已经把 captured 的 call_id 加入 captured_order(=新 tool_order),所以仍走有序路径。
2. captured_content + tool_bufs 互相缺失某 call_id(provider bug)。

只有 (2) 会真正命中字典序 fallback。字典序对 Anthropic 不重要(每个 tool_use block 独立),但对 OpenAI 兼容路径上的某些 strict provider(我记不清哪个)可能影响"tool_calls[]"数组顺序的稳定性 → 影响 cache prefix。

**建议**:加一行 `if !unordered_tool_calls.is_empty() { log::warn!("[byop] {n} tool_calls fell through to dict-sort fallback — provider inconsistency", n = unordered_tool_calls.len()) }`,既能在生产里发现 provider 异常,也避免 silent 走 fallback。

File: app/src/ai/agent_providers/chat_stream.rs [hunk 5]
```diff
@@ -2303,22 +2582,74 @@ pub async fn generate_byop_output(
...
+                    ));
+                }
+                _ => {}
+            }
         }
+        log::info!(
+            "[byop-diag] persistence summary: request_id={} task_id={} emitted_messages={} \
+             input_order={:?}",
+            request_id,
+            persistence_task_id,
```
Comment (zerx-lab) [inline app/src/ai/agent_providers/chat_stream.rs:2649]: **【正确性 — 高优先级】重构后丢失了 `referenced_attachments` 处理**

新实现里:
```rust
AIAgentInput::UserQuery { query, context, running_command, .. } => {
    let attachments = user_context::collect_user_attachments(context);
    // ... 持久化
    persistence_messages.push(make_user_query_message(
        persistence_task_id,
        &request_id,
        query.clone(),
        &attachments.binaries,
    ));
}
```

但 `build_chat_request` 的 UserQuery 分支(L817 区域)对当前轮 UserQuery 是这样组装的:
```rust
let mut prefixes: Vec<String> = Vec::new();
let user_attachments = user_context::collect_user_attachments(context);
if let Some(p) = &user_attachments.prefix {
    prefixes.push(p.clone());  // ← attachments.prefix(Block/SelectedText/File/Image 渲染的 XML)
}
if let Some(p) = user_context::render_referenced_attachments(referenced_attachments) {
    prefixes.push(p);  // ← @文件 / @block 引用
}
// ...full_text = prefixes + query + suffixes
```

而**持久化路径**只把 `binaries` 写进 `UserQuery.context.images`,**丢失了**:
1. `attachments.prefix`(Block / SelectedText 等结构化上下文)— 下一轮重建时 history UserQuery 分支(L640+)用 `render_api_referenced_attachments(&u.referenced_attachments)` 还原,但 `referenced_attachments` proto 字段在 `make_user_query_message` 里压根没填!
2. `referenced_attachments`(@提及)— 同上,持久化丢失。

**对比 main 分支**:旧实现的 `pending_user_queries: Vec<(String, Vec<UserBinary>)>` 与新实现完全等价,**这两个字段在持久化层一直就丢着**。所以 PR **没有引入新缺陷**,只是没顺手补上。

**结论**:这不算 PR 的回归 bug,但若 PR 描述声明"按 params.input 原顺序持久化 UserQuery/ActionResult" 暗示了完整保真,建议在 PR 描述里注明"binaries 之外的 attachment/referenced 上下文持久化仍走老路径,不在本 PR scope"。

**进一步**:如果以后要修这点,`make_user_query_message` 需要额外接收 `referenced_attachments: Vec<api::ReferencedAttachment>` 与 `attachments.prefix`,然后:
- referenced 写到 `UserQuery.referenced_attachments`(proto 字段已存在)
- prefix 写到 `UserQuery.context.executed_shell_commands` 之类(需研究 InputContext 哪个字段语义匹配)

或者退化方案:把 prefix + referenced 渲染好的字符串**直接拼到 query 文本前**再持久化,反正历史轮重建时也是 prepend 到 user message。两套路径合一,代价是历史里看到的 user query 文本与原始略有不同(多了 XML 块)。

File: app/src/ai/agent_providers/chat_stream.rs [hunk 6]
```diff
@@ -1033,6 +1033,307 @@ fn build_chat_request(
...
+                "[byop-diag] request orphan_or_misordered_tool_response idx={idx}: \
+                 response_call_ids={response_call_ids:?} previous_assistant_call_ids={previous_expected:?}"
+            );
+        }
+    }
+}
+
 /// 1:1 移植自 opencode `provider/transform.ts::applyCaching` 的 Anthropic 分支:
 /// 给 first 2 个 system message + last 2 个 non-system message 打 cache 标记。
 ///
```
Comment (zerx-lab) [inline app/src/ai/agent_providers/chat_stream.rs:1339]: **【可观测性 / 噪声】新增的 ~300 行 `log_chat_request_details` 日志量级过大**

每次请求都会 emit:
- 1 行 request summary(含 system_snippet 240 字符)
- N 行 request tool[idx](每个 tool 一行,典型 ~15-30 个内置工具 + MCP)
- 1 行 message_flow(整个对话所有 message 的 role+id 列表)
- M 行 request message[idx](每条 message 含 first_text 240 字符 snippet,tool_calls/responses 各带 args 240 字符 snippet)
- M 行 tool_pair / orphan 检测

一次普通的 5 轮对话(~30 条 message + 20 个 tool)就是 ~55 行 INFO log,**全部 240 字符 snippet**。

考虑到:
1. `app/src/ai/agent_providers/chat_stream.rs` 中**已有**许多 `[byop]` / `[byop-diag]` log;
2. 当前文件已有 `log::info!("[byop] diag_body_approx_len=...")` 和 `log::info!("[byop-diag] full_request_json=...")` 把**整个请求**塞进单行 log;
3. `[byop-diag] full_request_json` 已经覆盖所有诊断需求,新加的逐条 message 摘要属于重复。

**建议二选一**:
1. 把 `log_chat_request_details` 整体降级为 `log::debug!`,只在 `RUST_LOG=debug` 时输出;`full_request_json` 已能覆盖 INFO 级别诊断需求。
2. 保留为 INFO 但**移除** `first_text` / `args` / `system_snippet` 的实际内容(只保留 len / count),与 `full_request_json` 形成互补 —— 一个看结构、一个看完整内容。

当前实现两者都用 INFO + 都带 snippet,日志双倍冗余,生产环境(默认 log level INFO)磁盘/性能开销不必要。

**Review by zerx-lab — COMMENTED** [review]
# PR #96 Review 总结

## 整体评估：**修复方向正确,可合并,但建议先处理 2 个高优先级问题**

PR 修复的核心问题真实存在:用户在 pending tool 执行/取消前继续输入 query 时,`params.input` 顺序是 `[ActionResult, UserQuery]`,旧实现把"所有 UserQuery 先写,所有 ActionResult 后写",导致历史变成 `Assistant(tool_call) → UserQuery → ToolCallResult`,违反 Anthropic / OpenAI 的 `tool_use → tool_result` 邻接约束。

## 修复方案评估

### ✅ 合理的部分
1. **持久化按 `params.input` 原顺序写入** — 直接解决 root cause,实现简洁。
2. **`sanitize_tool_call_pairs` 升级到全局 call_id 收集** — 兜底机制,即使已有"被污染"历史也能修复。
3. **`tool_order` 跟踪与 captured/stream 合并** — 保留多 tool_call 在请求里的稳定顺序,符合 Anthropic 对 `tool_use` 块顺序的隐含期望。
4. **新增 2 个单元测试覆盖核心场景** — 行为正确性有保障。

### ⚠️ 高优先级关注点(建议合并前澄清/修复)

1. **[缓存命中] sanitize 跨位置重排会破坏 Anthropic prompt cache prefix 稳定性**
   - sanitize 现在把 Tool message 跨位置搬回 Assistant 后,改变了 messages 物理序;
   - 下一轮 `apply_caching_anthropic` 给"last 2 non-system"打的 5m breakpoint 会因相对位置漂移而 **必然 miss**;
   - 这与原 doc comment "**不做的事情:跨位置重排 Tool messages**" 直接冲突,doc 必须同步更新。
   - 建议加日志归因 + 在 cache 分析脚本里把此类 miss 单独列出。

2. **[正确性] End 处理把 `tool_bufs.entry(...).or_insert(...)` 改成 `tool_bufs.insert(...)`** —
   旧语义是"streaming 优先",新语义是"captured_content 覆盖" — 与注释表述一致,但属于行为变更,建议在 PR 描述里明确说明这是有意修改。

### 📋 中优先级关注点

3. **`responses_by_call_id` placeholder 覆盖逻辑** — 当前 match 语义正确但 4 路条件嵌套不直观;且**真实 result → 真实 result(后到覆盖)** 路径缺少单元测试覆盖。建议简化条件 + 加测试。

4. **流末 tool_calls fallback 字典序排序** — 仅在 provider 异常时命中,但**静默走 fallback** 可能掩盖问题。加一行 warn log 即可。

5. **`make_user_query_message` 仍然丢失 `referenced_attachments` 和 `attachments.prefix`** — 这是 main 分支早就存在的问题,**PR 没有引入回归**,但 PR 描述"按 params.input 原顺序持久化 UserQuery/ActionResult" 容易让人误以为完整保真。建议 PR 描述里注明 scope。

### 🔻 低优先级 / 噪声

6. **新增 ~300 行 `log_chat_request_details` 日志量过大** — 与已有 `[byop-diag] full_request_json` 重复诊断。建议整体降级为 `debug!`,或保留 INFO 但移除 snippet 内容。

## 建议合并路径

- **必做(阻塞合并)**:更新 `sanitize_tool_call_pairs` 的 doc comment(L1441-1448),明确声明现在确实做跨位置重排,并说明对 cache prefix 的潜在影响。
- **建议做(可后续 commit)**:End 处的 `insert` vs `or_insert` 行为变更在 PR 描述里加一句说明;`responses_by_call_id` 增加 real→real 覆盖测试。
- **可选优化**:日志降级、fallback warn、attachments 持久化补全可以单独 issue 跟踪。

整体上 PR **修复 root cause + 兜底 + 测试** 三件套齐全,代码质量良好,核心逻辑正确,**值得合并**;6 条 review 评论已逐点贴在对应行,请按优先级处理。

**Comment by zerx-lab** [comment]
已通过 commit [`bfca9ae`](https://github.com/zerx-lab/warp/commit/bfca9aec) 合并到 `main`,并在原 PR 改动基础上补了 review 里提到的几条改进:

- 更新 `sanitize_tool_call_pairs` doc:明确声明会跨位置重排 Tool messages,并解释对 Anthropic prompt cache prefix 稳定性的潜在影响(已知取舍)。
- 简化 `responses_by_call_id` 覆盖条件为更易读的 `should_replace` 写法,语义保持不变。
- 新增 `reordered_call_ids` 跟踪 + warn 日志,便于在 `[byop-cache]` miss 突增时关联归因。
- 流末 `dict-sort` fallback 加 warn:正常路径下 `tool_bufs` 应为空,命中说明 provider 在 ChunkArgs/captured_content 之间不一致。
- 新增 2 个回归测试:`real_tool_response_is_replaced_by_later_real_tool_response`、`placeholder_does_not_overwrite_existing_real_response`。

cargo check 通过,sanitize / cache_boundary / assistant_buffer 共 26 个单元测试全过。

原作者 @hemo94931 已通过 `Co-authored-by` 保留在 commit 中,感谢贡献!关闭该 PR。
